File: Backend/controllers/dataController.js
const fs = require('fs');
const path = require('path');
const Data = require('../models/dataModel');
const { validateWaterQualityData } = require('../utils/validation');
const { calculateRiskScore, getMitigationRecommendation } = require('../utils/helpers');

const DATA_MASUK_DIR = path.join(__dirname, '..', '..', 'data_masuk');

const appendToCsv = (row) => {
  try {
    if (!fs.existsSync(DATA_MASUK_DIR)) fs.mkdirSync(DATA_MASUK_DIR, { recursive: true });
    const file = path.join(DATA_MASUK_DIR, `sensor_${new Date().toISOString().slice(0, 10)}.csv`);
    if (!fs.existsSync(file)) {
      fs.writeFileSync(file, 'id,tanggal,lokasi,suhu,pH,salinitas,kekeruhan\n');
    }
    fs.appendFileSync(file, `${row.id},${row.tanggal},${row.lokasi},${row.suhu},${row.pH},${row.salinitas},${row.kekeruhan}\n`);
  } catch (err) {
    console.error('Gagal menulis CSV data_masuk:', err.message);
  }
};

const DataController = {
  getAll: async (req, res) => {
    try {
      const { page = 1, limit = 50 } = req.query;
      const offset = (page - 1) * limit;
      const data = await Data.findAll(limit, offset, req.user.id);
      const total = await Data.count(req.user.id);
      res.json({ success: true, data, pagination: { page: parseInt(page), limit: parseInt(limit), total, pages: Math.ceil(total / limit) } });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  getLatest: async (req, res) => {
    try {
      const data = await Data.findLatest(req.user.id);
      if (!data) return res.status(404).json({ success: false, message: 'Data tidak ditemukan' });
      const risk = calculateRiskScore(data.suhu, data.pH, data.salinitas, data.kekeruhan);
      const rekomendasi = getMitigationRecommendation(risk.status);
      res.json({ success: true, data: { ...data, skor_risiko: risk.skor, status: risk.status, rekomendasi } });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  getLastDay: async (req, res) => {
    try {
      const data = await Data.findLastDay(req.user.id);
      const enriched = data.map(d => {
        const risk = calculateRiskScore(d.suhu, d.pH, d.salinitas, d.kekeruhan);
        return { ...d, skor_risiko: risk.skor, status: risk.status };
      });
      res.json({ success: true, data: enriched });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  getLatestPerLokasi: async (req, res) => {
    try {
      const data = await Data.findLastDay(req.user.id);
      const worst = {};
      data.forEach((d) => {
        const risk = calculateRiskScore(d.suhu, d.pH, d.salinitas, d.kekeruhan);
        const rekomendasi = getMitigationRecommendation(risk.status);
        const enriched = { ...d, skor_risiko: risk.skor, status: risk.status, rekomendasi };
        if (!worst[d.lokasi] || enriched.skor_risiko < worst[d.lokasi].skor_risiko) {
          worst[d.lokasi] = enriched;
        }
      });
      res.json({ success: true, data: Object.values(worst) });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  getById: async (req, res) => {
    try {
      const data = await Data.findById(req.params.id, req.user.id);
      if (!data) return res.status(404).json({ success: false, message: 'Data tidak ditemukan' });
      const risk = calculateRiskScore(data.suhu, data.pH, data.salinitas, data.kekeruhan);
      res.json({ success: true, data: { ...data, skor_risiko: risk.skor, status: risk.status } });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  create: async (req, res) => {
    try {
      const errors = validateWaterQualityData(req.body);
      if (errors.length) return res.status(400).json({ success: false, message: errors.join(', ') });
      const id = await Data.create({ ...req.body, user_id: req.user.id });
      const newData = await Data.findById(id, req.user.id);
      appendToCsv(newData);
      res.status(201).json({ success: true, data: newData });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  update: async (req, res) => {
    try {
      const errors = validateWaterQualityData(req.body);
      if (errors.length) return res.status(400).json({ success: false, message: errors.join(', ') });
      const affected = await Data.update(req.params.id, req.body, req.user.id);
      if (!affected) return res.status(404).json({ success: false, message: 'Data tidak ditemukan' });
      const updated = await Data.findById(req.params.id, req.user.id);
      res.json({ success: true, data: updated });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  delete: async (req, res) => {
    try {
      const affected = await Data.delete(req.params.id, req.user.id);
      if (!affected) return res.status(404).json({ success: false, message: 'Data tidak ditemukan' });
      res.json({ success: true, message: 'Data dihapus' });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  getStats: async (req, res) => {
    try {
      const stats = await Data.getStats(req.user.id);
      res.json({ success: true, data: stats });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  }
};

module.exports = DataController;

File: backend_python/controllers/data_controller.py
import csv
import math
import os
from datetime import datetime

from flask import g, request

from models.data_model import Data
from utils.helpers import calculate_risk_score, fail, get_mitigation_recommendation, jsonable, ok
from utils.validation import validate_water_quality_data

DATA_MASUK_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data_masuk")


def append_to_csv(row):
    try:
        os.makedirs(DATA_MASUK_DIR, exist_ok=True)
        path = os.path.join(DATA_MASUK_DIR, f"sensor_{datetime.now():%Y-%m-%d}.csv")
        baru = not os.path.exists(path)
        with open(path, "a", newline="", encoding="utf-8") as f:
            writer = csv.writer(f)
            if baru:
                writer.writerow(["id", "tanggal", "lokasi", "suhu", "pH", "salinitas", "kekeruhan"])
            writer.writerow(
                [row["id"], row["tanggal"], row["lokasi"], row["suhu"], row["pH"], row["salinitas"], row["kekeruhan"]]
            )
    except OSError as e:
        print(f"Gagal menulis CSV data_masuk: {e}")


def enrich(row):
    risk = calculate_risk_score(row["suhu"], row["pH"], row["salinitas"], row["kekeruhan"])
    return {**row, "skor_risiko": risk["skor"], "status": risk["status"]}


class DataController:
    @staticmethod
    def get_all():
        page = request.args.get("page", 1, type=int)
        limit = request.args.get("limit", 50, type=int)
        offset = (page - 1) * limit
        rows = Data.find_all(limit, offset, g.user["id"])
        total = Data.count(g.user["id"])
        return ok(
            {
                "success": True,
                "data": rows,
                "pagination": {
                    "page": page,
                    "limit": limit,
                    "total": total,
                    "pages": math.ceil(total / limit) if limit else 0,
                },
            }
        )

    @staticmethod
    def get_latest():
        row = Data.find_latest(g.user["id"])
        if not row:
            return fail("Data tidak ditemukan", 404)
        risk = calculate_risk_score(row["suhu"], row["pH"], row["salinitas"], row["kekeruhan"])
        rekomendasi = get_mitigation_recommendation(risk["status"])
        return ok({"success": True, "data": {**row, "skor_risiko": risk["skor"], "status": risk["status"], "rekomendasi": rekomendasi}})

    @staticmethod
    def get_last_day():
        rows = Data.find_last_day(g.user["id"])
        enriched = []
        for r in rows:
            risk = calculate_risk_score(r["suhu"], r["pH"], r["salinitas"], r["kekeruhan"])
            enriched.append({**r, "skor_risiko": risk["skor"], "status": risk["status"]})
        return ok({"success": True, "data": enriched})

    @staticmethod
    def get_latest_per_lokasi():
        rows = Data.find_last_day(g.user["id"])
        worst = {}
        for r in rows:
            risk = calculate_risk_score(r["suhu"], r["pH"], r["salinitas"], r["kekeruhan"])
            rekomendasi = get_mitigation_recommendation(risk["status"])
            item = {**r, "skor_risiko": risk["skor"], "status": risk["status"], "rekomendasi": rekomendasi}
            if r["lokasi"] not in worst or item["skor_risiko"] < worst[r["lokasi"]]["skor_risiko"]:
                worst[r["lokasi"]] = item
        return ok({"success": True, "data": list(worst.values())})

    @staticmethod
    def get_by_id(row_id):
        row = Data.find_by_id(row_id, g.user["id"])
        if not row:
            return fail("Data tidak ditemukan", 404)
        risk = calculate_risk_score(row["suhu"], row["pH"], row["salinitas"], row["kekeruhan"])
        return ok({"success": True, "data": {**row, "skor_risiko": risk["skor"], "status": risk["status"]}})

    @staticmethod
    def create():
        body = request.get_json(silent=True) or {}
        errors = validate_water_quality_data(body)
        if errors:
            return fail(", ".join(errors), 400)
        new_id = Data.create({**body, "user_id": g.user["id"]})
        new_row = Data.find_by_id(new_id, g.user["id"])
        append_to_csv(new_row)
        return ok({"success": True, "data": new_row}, 201)

    @staticmethod
    def update(row_id):
        body = request.get_json(silent=True) or {}
        errors = validate_water_quality_data(body)
        if errors:
            return fail(", ".join(errors), 400)
        affected = Data.update(row_id, body, g.user["id"])
        if not affected:
            return fail("Data tidak ditemukan", 404)
        return ok({"success": True, "data": Data.find_by_id(row_id, g.user["id"])})

    @staticmethod
    def delete_row(row_id):
        affected = Data.delete(row_id, g.user["id"])
        if not affected:
            return fail("Data tidak ditemukan", 404)
        return ok({"success": True, "message": "Data dihapus"})

    @staticmethod
    def get_stats():
        stats = Data.get_stats(g.user["id"])
        return ok({"success": True, "data": stats})
